example.js

```diff
@@ -7,6 +7,13 @@ const app = createApp({
     unleashApiToken: '56907a2fa53c1d16101d509a10b78e36190b0f918d9f122d',
     clientKeys: ['proxy-secret', 'another-proxy-secret', 's1'],
     refreshInterval: 1000,
+    bootstrapTokens: ['bootstrap-token'],
+    bootstrap: {
+        url: 'https://localhost:4000',
+        urlHeaders: {
+            Authorization: 'some-token',
+        },
+    },
```

Comment:
What are the options here? What are they used for? Maybe it's because I haven't looked that closely into the bootstrap update in the node sdk, but I don't understand what this configuration is. Is this to bootstrap the proxy or is it to support bootstrap in clients?

I've got a hunch it's about bootstrapping the proxy itself, but have some questions:

- `bootstrapTokens`: what's this used for? 

- `bootstrap.url`: Is this where the proxy fetches the configuration from? what should the url point to? An unleash instance? A different proxy? 

- `bootstrap.urlHeaders`: I'm assuming these headers just get passed directly to the request performed to the `url` above, so the `authorization` prop should be valid for that url.

Reply:
Perfect questions, and why I included you early. For clearity I try to cover to use cases: 

1. Allow the Unleash-Proxy to bootstrap from somewhere. Could be another Unleash-Proxy instance, a S3 bucket, or wherever you have a fresh feature-toggle-configuration state stored. 
 - `bootstrap.url` - where the proxy should bootstrap from
 - `bootstrap.urlHeaders` - headers to be added to the request. Typically the `Authorization` header. 
2. Allow any server-side SDK to bootstrap from the the Unleash Proxy. 
  - The SDKs should connect to the built in "bootstrap url" and provide a "boostrapToken". This needs to be differnt than the `clientKeys`, as the boostrap data return will be the full local cache of all feature toggle configuration. 

I am really unsure on the naming of things.

Reply:
I have prefixed all bootstrap options with `exp*` to allow us to release this as part of the proxy and start testing this feature behind the scenes.

Reply:
Thanks for the clarification! I think I understand it better now. I've got a few followup questions and suggestions if you don't mind.

First: When the proxy bootstraps off something else, you'll need to provide a bootstrap url and (optional) headers. When a client wants to bootstrap off the proxy, the client will need to provide one of the specified bootstrap tokens. Is this the correct understanding?

## Regarding bootstrap tokens:

1. So the bootstrap tokens are more akin to regular API tokens (such as the ones distributed by Unleash now), right? And as with Unleash proper, they should not be used by client-side SDKs. 
2. Does/will the proxy give CORS errors too if you try and access the bootstrap data from a client SDK?
3. When using the bootstrap token, does it go into the `authorization` header? Will we require/accept a scheme for the auth header in that case, e.g. `authorization: bearer <token>`?

## Thoughts on naming

Because `bootstrapTokens` has a very similar name to the `bootstrap` object, it's easy to assume that they're part of the same configuration. While both are used for bootstrapping, one is for bootstrapping the proxy and the other is for using the proxy as a bootstrap host (server? source?). Can we somehow distinguish between these better? 

I don't know what would be a good naming scheme, but I can imagine them being two separate objects, such as `bootstrapSourceConfig` and `bootstrapClientConfig` or something. `bootstrapProvider` and `bootstrapConsumer`, perhaps? Or `bootstrap.provider` and `bootstrap.consumer` / `bootstrap.asProvider` and `bootstrap.asConsumer`. Hmm 🤔

Reply:
I am thinking maybe we can call the `bootstrapTokens` something else. It actually supports the SDK to also use that endpoint as their Unleash API endpoint. Maybe we can just call it `serverSdkKeys` or `serverKeys`?

Reply:
> Does/will the proxy give CORS errors too if you try and access the bootstrap data from a client SDK?

This does not make sense. The client SDK connects to the proxy and bootstrapping from the proxy would not be needed. Client SDKs only support local boostrapping.

Reply:
> I am thinking maybe we can call the bootstrapTokens something else. It actually supports the SDK to also use that endpoint as their Unleash API endpoint. Maybe we can just call it serverSdkKeys or serverKeys?

Yeah, that's a good point. It'd be nice to call it `apiKeys`, but I'm afraid that'd be really confusing (when you've already got Unleash API tokens). `proxyApiKeys`, maybe? Still confusing, though.  

> This does not make sense. The client SDK connects to the proxy and bootstrapping from the proxy would not be needed. Client SDKs only support local boostrapping.

Regardless of whether it makes sense or not, will we allow it? Just thinking that if it doesn't give a hard error (such as CORS), then someone's gonna do it to get all the toggle data as a way of working around something else. If I understand correctly, then the bootstrap data contains all the data the server-side SDKs need to get going, so we still don't want to make that available to client-side SDKs, right?

Reply:
> > I am thinking maybe we can call the bootstrapTokens something else. It actually supports the SDK to also use that endpoint as their Unleash API endpoint. Maybe we can just call it serverSdkKeys or serverKeys?
> 
> Yeah, that's a good point. It'd be nice to call it `apiKeys`, but I'm afraid that'd be really confusing (when you've already got Unleash API tokens). `proxyApiKeys`, maybe? Still confusing, though.
> 
> > This does not make sense. The client SDK connects to the proxy and bootstrapping from the proxy would not be needed. Client SDKs only support local boostrapping.
> 
> Regardless of whether it makes sense or not, will we allow it? Just thinking that if it doesn't give a hard error (such as CORS), then someone's gonna do it to get all the toggle data as a way of working around something else. If I understand correctly, then the bootstrap data contains all the data the server-side SDKs need to get going, so we still don't want to make that available to client-side SDKs, right?

It would be really hard to outright disallow it. We could of course block it with CORS, but this would only apply to the browser. You could still make the HTTP call with the correct token and receive the data through any HTTP client that does not implement CORS. I do think it's worth blocking with CORS to avoid unintentional usage of this through the client side SDKs, even if we can't outright disallow it.

Reply:
@FredrikOseberg Yeah, that's what I meant 😅 Sorry if I wasn't clear on it. But yeah, cors blocking it makes sense to me 🙋🏼